### PR TITLE
[FIX] sale_timesheet: filter task SO without commercial partner id

### PR DESCRIPTION
**Pull request by Yoann-bary**
Steps to reproduce:
- Project > Pick any task > Debug Mode
- Studio > View tab > Tick 'Show invisible elements'
- Scroll to 'Sales Order' and click it
- Untick 'Invisible' then tick and untick readonly
- Close > you should have a Sales Order field on the task
- Click to change it's value

An error occurs because the 'commercial_partner_id' field was removed from task in 17.0 in dcbdb6e690f29bc5327d7067688c93071d9a6b2d. Because of this the domain which filters 'Sales Order' (which contains this field) cannot be evaluated.

Since the field is still available on the sale_order model, we can simply invert the child_of relation:
from sale_order.partner_id child_of task.commercial_partner_id to sale_order.commercial_partner_id parent_of task.partner_id Which should serve essentially the same purpose.

opw-4199947

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/sale_timesheet/wizard/project_create_sale_order.py [hunk 1]
```diff
@@ -50,7 +50,7 @@ def default_get(self, fields):
 
     sale_order_id = fields.Many2one(
         'sale.order', string="Sales Order",
-        domain="['|', '|', ('partner_id', '=', partner_id), ('partner_id', 'child_of', commercial_partner_id), ('partner_id', 'parent_of', partner_id)]")
```
Comment (nle-odoo) [inline addons/sale_timesheet/wizard/project_create_sale_order.py:53]: commercial_partner_id is 2 lines above this, is this part of the change necessary?

File: addons/sale_timesheet/models/project.py [hunk 2]
```diff
@@ -537,7 +537,7 @@ def _get_default_partner_id(self, project, parent):
                 return related_project.sale_line_employee_ids.sale_line_id.order_partner_id[:1]
         return res
 
-    sale_order_id = fields.Many2one(domain="['|', '|', ('partner_id', '=', partner_id), ('partner_id', 'child_of', commercial_partner_id), ('partner_id', 'parent_of', partner_id)]")
+    sale_order_id = fields.Many2one(domain="['|', ('partner_id', '=', partner_id), ('commercial_partner_id', 'parent_of', partner_id)]")
```
Comment (nle-odoo) [inline addons/sale_timesheet/models/project.py:540]: isn't `('partner_id', 'parent_of', partner_id)` less restrictive than `('commercial_partner_id', 'parent_of', partner_id)`?

if we have eg.

- company A:
-- company B
--- has child partner C
---- which has child partner D

then if the partner_id is the current record is D, I get:

- with `('commercial_partner_id', 'parent_of', partner_id)` => partner D, company B
- with `('partner_id', 'parent_of', partner_id)` => partner D, partner C, company B, company A

also it doesn't seems like we can use `commercial_partner_id` since it's added by sale_subscription which is not a dependency of sale_timesheet:

https://github.com/odoo/enterprise/blob/071d19332b5bb6cdb96dafd64fe463e2ed0eef49/sale_subscription/models/sale_order.py#L105

we could use `('partner_id.commercial_partner_id', 'parent_of', partner_id)` but this doesn't solve the issue that what we wanted was `('partner_id', 'child_of', partner_id.commercial_partner_id)`

**Review by nle-odoo — APPROVED** [review]
seems good to me to remove "('partner_id', 'child_of', commercial_partner_id)" if there is no more commercial_partner_id on the model

if we want to have it back I don't see a solution in stable. if it was really necessary we could:

- add a related non-stored field "commercial_partner_id"
- add this field in the view (it would be automatically added in 18.0 I think)

this would mean that people that don't update the view would still have an error but it will be solved if someone update the view or is on a new instance (and they were getting an error before)

To not have an error, we could:

- remove `('partner_id', 'child_of', commercial_partner_id)` from the python field domain
- add the domain with `('partner_id', 'child_of', commercial_partner_id)` in the XML view field domain

This way people that don't update the view should not have the error. And people that update the view should have the domain applied.

TL;DR But since this doesn't seem to been reported since 2023 removal of the field, and is needing studio to reproduce, your current solution of only removing "('partner_id', 'child_of', commercial_partner_id)" is probably enough. I don't think we need to do too much effort to solve this if it's unused code.

File: addons/sale_timesheet/models/project.py (hunk 2 above)
Comment (Yoann-bary) [reply addons/sale_timesheet/models/project.py:540]: Based on https://github.com/odoo/odoo/blob/17.0/odoo/addons/base/models/res_partner.py#L428-L434 I think it would be:
- `('commercial_partner_id', 'parent_of', partner_id)` => partner D, partner C, company B
But you're right, it doesn't include company A which it should.

Also, I thought that `('partner_id.commercial_partner_id', 'parent_of', partner_id)` and `('partner_id', 'child_of', partner_id.commercial_partner_id)` were equivalent, as in it only checks whether both partners have  a parent in common. Am I misunderstanding something ?

|                                                    |    |      |    |     |
| ---------------------------------------- | --- | -- | -- | -- |
| task.partner_id                        | A | B | C | D |
| task.commercial_partner_id | A | B | B | B |
| so.partner_id                           | A | B | C | D |
| so.commercial_partner_id    | A | B | B | B |

Comment (nle-odoo) [reply addons/sale_timesheet/models/project.py]: I could have thought that also (and it's what I thought at first).  But testing it doesn't seem to be the case.

With my example and checking step by step in odoo/osv/expression.py:

company A:
-- company B
--- has child partner C
---- which has child partner D
---- which has child partner E
company F

('partner_id.commercial_partner_id', 'parent_of', partner_id)

When we have partner_id that is "partner D" we will:

1) this will search `('partner_id', 'any', [('commercial_partner_id', 'parent_of', <partner D>)])`
2) this will search on res.partner [('commercial_partner_id', 'parent_of', <partner D>)]
3) this will return res.partner that are commercial_partner_id of partner_id and itself => "partner D", "company B"
4) then inserting it in the original query we will have `('partner_id', 'in', [<partner D>, <company B>])])`
=> which gives us just the same order with the partner being the given partner or exactly the parent

While `('partner_id', 'child_of', partner_id.commercial_partner_id)` gives us all the sale order id of a partner that is a children of the commercial partner ID.

But after checking understand this in more detail, it seems that what we want would be possible with:

```py
('partner_id.commercial_partner_id.id', 'parent_of', <partner D>)
```
In this case, we will:

- first search the res.partner parent of <partner D> -> "company A", "company B" "partner D", "company B"
- then search the res.partner that have a commercial_partner_id in these results (all besides company F)
- and search the sale.order with a partner_id matching partner_id besides company F

**Review by nle-odoo — APPROVED** [review]
seems good to me with `('partner_id.commercial_partner_id.id', 'parent_of', partner_id)`

or would also seems good if we don't want that (because it is a little larger, since it will also contain the parent of commercial_id) to just remove it since this field doesn't seem used a lot (not sure if it can be shown in the interface with odoo normal code)

File: addons/sale_timesheet/models/project.py [hunk 3]
```diff
@@ -537,7 +537,7 @@ def _get_default_partner_id(self, project, parent):
                 return related_project.sale_line_employee_ids.sale_line_id.order_partner_id[:1]
         return res
 
-    sale_order_id = fields.Many2one(domain="['|', '|', ('partner_id', '=', partner_id), ('partner_id', 'child_of', commercial_partner_id), ('partner_id', 'parent_of', partner_id)]")
+    sale_order_id = fields.Many2one(domain="['|', '|', ('partner_id', '=', partner_id), ('partner_id.commercial_partner_id.id', 'parent_of', partner_id), ('partner_id', 'parent_of', partner_id)]")
```
Comment (xavierbol) [inline addons/sale_timesheet/models/project.py:540]: `.id` is really needed? If yes, why?

Does it not equivalent to:

```suggestion
    sale_order_id = fields.Many2one(domain="['|', '|', ('partner_id', '=', partner_id), ('partner_id.commercial_partner_id', 'parent_of', partner_id), ('partner_id', 'parent_of', partner_id)]")
```
? 🤔

**Review by xavierbol — COMMENTED** [review]
Thanks for your work, I left one comment. 🙂

File: addons/sale_timesheet/models/project.py (hunk 3 above)
Comment (xavierbol) [reply addons/sale_timesheet/models/project.py:540]: Wow, I did not know that 🤯 

Thanks for the explanation! 🙂 

Then LGTM! 💪🏼

**Comment by xavierbol** [comment]
robodoo r+

**Comment by xavierbol** [comment]
robodoo retry
